controller/controller_server/csi_controller_server.py

```diff
@@ -534,6 +539,91 @@ def GetCapacity(self, request, context):
         logger.info("finished GetCapacity")
         return csi_pb2.GetCapacityResponse()
 
+    def ControllerExpandVolume(self, request, context):
+        set_current_thread_name(request.volume_id)
+        logger.info("ControllerExpandVolume")
+        secrets = request.secrets
+
+        try:
+            utils.validate_expand_volume_request(request)
+
+            user, password, array_addresses = utils.get_array_connection_info_from_secret(secrets)
+
+            try:
+                array_type, volume_id = utils.get_volume_id_info(request.volume_id)
+            except ObjectIdError as ex:
+                logger.warning("volume id is invalid. error : {}".format(ex))
+                context.set_code(grpc.StatusCode.INVALID_ARGUMENT)
+                return csi_pb2.ControllerExpandVolumeResponse()
+
+            with get_agent(user, password, array_addresses, array_type).get_mediator() as array_mediator:
+                logger.debug(array_mediator)
+
+                required_bytes = request.capacity_range.required_bytes
+                min_size = array_mediator.minimal_volume_size_in_bytes
+                max_size = array_mediator.maximal_volume_size_in_bytes
+
+                volume_before_expand = array_mediator.get_object_by_id(volume_id, config.VOLUME_TYPE_NAME)
+                if not volume_before_expand:
+                    raise controller_errors.ObjectNotFoundError(volume_id)
+
+                if required_bytes == 0:
+                    context.set_code(grpc.StatusCode.OK)
+                    return csi_pb2.ControllerExpandVolumeResponse()
+
+                if not min_size <= required_bytes <= max_size:
```

Comment:
now that I think about it - I think it's ok for `required_bytes` to be smaller than `min_size`.
what do you think?

Reply:
I think that the [spec](https://github.com/container-storage-interface/spec/blob/v1.3.0/spec.md#controllerexpandvolume-errors) expect in out_of_range that the capacity_range will be inside our limits

Reply:
could you elaborate?

here are the details of my interpretation:
`If a volume corresponding to the specified volume ID is already larger than or equal to the target capacity of the expansion request, the plugin SHOULD reply 0 OK.`

`OUT_OF_RANGE Indicates that the capacity` **range** `is not allowed by the Plugin.`

`CapacityRange.required_bytes`: `Volume MUST be` **at least** `this big.`

Reply:
as I see it - `CapacityRange` contains the range of the new volume size , `required_bytes` is min and `limit_bytes` is max. in `OUT_OF_RANGE` we need to check the range given by the request- the min and max. so  if `required_bytes `(or "the min") is lower then our min limitation, we suppose to return `OUT_OF_RANGE `..

Reply:
but why would they expect us to enforce the *entire* requested range?
if someone asks you for any value that is larger than 1, and you can provide values only between 3 and 5, why wouldn't you just give them e.g. 3?

Reply:
similar question - if they only specify `limit_bytes` of 10, and our `maximal_volume_size_in_bytes` is only 9, do you think we should fail with `OUT_OF_RANGE` as well?
I think it just means that the request is more "flexible" than us, which should increase the likelihood for a success.

Reply:
well, yes, but in what size the costumer actually wanted the volume?

Reply:
`To specify an exact size, required_bytes and limit_bytes SHALL be set to the same value.`
(otherwise, the wanted size is a range, not an exact value)

Reply:
@erantzabari can we have a third opinion?

Reply:
Starting from the bottom line - We don't need to fire out OUT_OF_RANGE in case required_bytes is smaller than min_size.
The main goal (also as far as I read in the spec) is that we want to expend volume which is size is smaller than the requested size. If the volume size is larger than the requested size, we shall return OK neglected to boundaries.

Reply:
done